Support for testing classes with non-public ctors

appakz opened the pull request:
Howdy SpecEasy-ians.

We use a fair bit of internal classes in our projects and often want to have those classes use internal constructors. We've been having issues with the TinyIoC implementation we have embedded into SpecyEasy resolving and creating those SUT instances if the target type has an internal ctor. The issue was that TinyIoC wasn't finding the internal ctors via the reflection calls it was making and therefore just blew  up while trying to build the SUT instance.

I created an extra class lib in the solution (with the Dumbest Code Ever(tm) in it) with 'InternalsVisibleTo' assembly meta data to repro the issue. As it turns out, the fix appears to be pretty easy. There was some commented out code in the TinyIoC file that was using the newer 'DeclaredConstructors' property that was made available in .NET 4.5, and that property returns ''all'' constructors (not just public ones).  If I were building a general purpose IoC container I'd say that behavior is maybe not always desired, but for the purposes of SpecEasy I'd say that we probably do want to allow for constructing SUT instances using non-public ctors if possible.

A couple of other notes:
- This change would mean that TinyIoC could resolve SUTs that have private constructors. I'm not sure how I feel about that exactly... I'm open to thoughts or concerns though.
- There was some weirdness with the nUnit package when I first cloned the repo and tried to build it, I was missing refs to nunit.core and nunit.core.interfaces. The .csproj file wanted them to be under /packages/nunit.test.runner (or something like that), but our packages config didn't include that page. I went ahead and included it, and that made things work, but I noticed also bumped the version of those assemblies from 2.6.3.<something> to 2.6.4.<something>. This is only for the SpecEasy.Specs assembly and won't actually impact the Speceasy.dll that we're shipping, so it shouldn't matter but I wanted to mention it.
- I made this change with CTP of Visual Studio 2015, and it tagged the .sln file accordingly. Not sure that we care.
- The SpecEasy.Specs project got a new "service include={guid}" node added that apparently is harmless and just used by the built-in VS test runner to know that a certain project contains unit tests.

Let me know what you think!

appakz commented:
ping @rockhymas @ronrat

dbertram commented on SpecEasy.Specs/ExternalSpec/MagicNumberAdderSpec.cs:
```diff
@@ -0,0 +1,24 @@
+            When("adding two numbers together", () => result = SUT.AddNumbers(a, b));
+
+            Given("the first number is 3", () => a = 3).Verify(() =>
+                Given("the magic numbe associated with 3 is 7", () => Get<IMagicNumberLookup>().Stub(l => l.Lookup(3)).Return(7)).Verify(() =>
```
Super minor, but there's a typo in the `Given` description here: "the magic numbe**r**..."

appakz replied:
This is where I wish I could just edit the file here in the diff....

dbertram commented on SpecEasy/TinyIoC.cs:
```diff
@@ -3619,12 +3619,8 @@ private ConstructorInfo GetBestConstructor(Type type, NamedParameterOverloads pa
-            return type.GetConstructors().OrderByDescending(ctor => ctor.GetParameters().Count());
-//#endif
-        }
+			return type.GetTypeInfo().DeclaredConstructors.OrderByDescending(ctor => ctor.GetParameters().Count());
```
Indentation is off by one here I think.

dbertram commented:
@appakz How does constructor precedence work with this change? If I have a type I'm speccing that has public, internal, and private constructors, should we have a primary and secondary sort maybe? I.e., primary sort would be visibility, secondary sort would be the number of constructor parameters?

The current code just uses the number of constructor parameters--which I think made sense when it was only dealing with public constructors--but I feel like the expected behavior might be to favor public over internal and internal over private. What do you think?

Related to the concern about using private constructors, we could also just filter those out from what's returned from `GetTypeConstructors`, no?

appakz commented:
@dbertram that's an interesting idea about ordering. The old code using Type.GetConstructors would have only ever returned public constructors ordered by parameter count, and I'm not sure there's any guarantee that the public ones are coming back first from the DeclaredConstructors property, so it'd probably be a good idea to sort the ConstructorInfo instances coming back from GetTypeConstructors to provide the public ones first, followed by internal, followed by protected, followed by private maybe? That would help ensure that any existing specs out there with SUTs that had public and non-public ctors would continue to find the same public ctor that they are currently after this change is made. I'll make that fix before we merge this (if we merge it).

We can filter our private ctors pretty easily as well if we care to. There are lots of properties on ConstructorInfo ('IsPrivate' namely) that will let us do that. I'm not sure I care enough to at this point though...

On a somewhat related note, I've been considering adding the ability to just replace the method by which we build SUT instances completely. We can do this now by making sure that the first reference to the SUT that executes in the spec is "SUT = new <whatever>..", but I was thinking about making that something that is more explicitly supported by speceasy.

dbertram commented:
@appakz Personally, I think filtering out private constructors is reasonable, but I leave it to you for whether it's worth including here or not. Certainly straightforward to add later if it causes problems.

And I agree wrt having a better means for explicitly choosing which constructor for SUT is used within a spec. Maybe something like a `WithConstructor` call that registers a lambda responsible for creating SUT? That way you could use a different constructor for different specs within the same class (versus something overridable like `BeforeEachExample` that's class-level).  Or did you have something cooler in mind? ;)

appakz commented:
@dbertram some kind of way to provide a lambda to construct the SUT is what I was thinking.... not sure exactly how that looks yet though.

rockhymas commented:
On the question of ordering, we should at the very least order all public constructors before any others, so that when you upgrade speceasy to a version with this change you don't have a change in behavior.

rockhymas commented:
I agree with the idea that we should not include private constructors.

ronrat commented:
+1 for ordering ctors such that public comes first so that existing tests are not inadvertently affected. As far as the ordering of the the other access modifiers, I'm ok w/ the order that @appakz suggested. 

I'm somewhat ambivalent about excluding private constructors but I don't remember a use case where using a private constructor would have been helpful so I vote to exclude them.

I really like the idea of providing a way to override the default SUT construction. I often run into scenarios where I have to do exactly what @appakz said: construct the SUT using the `SUT = new ...` before anything accesses that property. It works but I don't think it is obvious to new users of SpecEasy that it's how you get around that issue. Providing a documented way to override the SUT creation would really help clear that up.

appakz commented:
Ok. I'm going to grab the newly released VS 2015 so I can make these changes using it instead of the CTP.

appakz commented:
Sorry I took so long to get back to this. I was on vacation for awhile but I think I've circled back and done everything that we discussed doing. I'm tired and was away from this code for awhile so I'm relying on you all to spot the silly mistakes that might be lurking in here. I'll take a look again myself with fresh eyes tomorrow.

I discovered while working on this that the existing sort based on parameter counts actually doesn't work the way I thought it did. There are two means by which TinyIoC can try to resolve ctors, but since we use such a limited feature set we only use one of them. The one that we use will favor *lower* parameter counts over higher ones. I'm not sure that's the right thing to do, but changing it now would technically be a breaking change and I didn't want to take it on. I'd like to add a nicer way to control how the SUTs are constructed, and at that point maybe rip some of the stuff that we don't use or like out of our embedded TinyIoC and make a major version rev with the breaking change on constructor count sorting.

I also added some tests specifically to illustrate how TinyIoC will pick constructors mostly so I could validate the work I was doing. To facilitate these I also added a 'EnsureSUT' method in the base Spec<T> that will just make sure that the SUT instance has been created via TinyIoC. I've wanted something like this in the past and have ended up adding my own little hacks to the specs I was writing to accomplish this.

appakz commented:
@dbertram Added a new unit test to assert that we never use a private ctor, and added the named constants.

dbertram commented:
@appakz Sweet!

Should those `const`s be all caps to match the convention used elsewhere?

rockhymas commented:
I like the EnsureSUT method, that will be nice to have.

rockhymas commented:
As for constructor parameter count, I'm not sure I have a preference on ordering lowest first or highest first, but I do think that multiple constructors in itself could be a code smell. And on the private constructors, I still feel they should be excluded, so I'm glad to see we have tests verifying that. If a class has a private constructor, it almost always has other ways of getting instances (i.e. a factory), which should be what is tested.

rockhymas commented:
And overall, this change looks good to me.

appakz commented:
Renamed the constants for consistency. Unless anyone has any other objections, could we merge this PR in? (I don't have write access to do so). Who does? @Fammy @ronrat  @mmertsock  ?

I will have another PR coming soon after this one is merged....
